include/curl/easy.h

```diff
@@ -25,6 +25,98 @@
 extern "C" {
 #endif
 
+/* Application can use a memory PKCS12 certificate with CURLOPT_SSLCERT
+ *  with OpenSSL, Schannel or SecTransport
+ *
+ *
+ *  Example if with certificate binary data and size in:
+ *    void* certdata;
+ *    size_t certsize;
+ *    ...
+ *  set the certificate with:
+ *    struct curl_blob structblob;
+ *    curl_init_blob_dup(&structblob, certdata, certsize);
+ *    curl_easy_setopt(curl, CURLOPT_KEYPASSWD, &structblob);
+ *             my_setopt_str(curl, CURLOPT_SSLCERT, &structblob);
+ *  struct curl_blob is just a 0x20 bytes structures which will
+ *  start bt a magic string and contain data
+ *
+ *  with curl_init_blob_dup, certdata can be discarded after
+ *    calling curl_easy_setopt
+ *  with curl_init_blob_persist, certdata need be valid when the
+ *  connexion is made
+ *
+ * Note : never call curl_init_blob_xxx without a pointer to a
+ *   valid 0x20 bytes block
+ * And never use the string CURL_BLOB_MAGIC without a correct
+ *   blob block
+ *
+ *
+ * about the idea, see https://curl.haxx.se/mail/lib-2016-09/0074.html */
+
+#define CURL_BLOB_MAGIC          ("\x01" "CurlMemBlob" "\xff\x01")
+#define CURL_BLOB_MAGIC_STRLEN   (14)
+#define CURL_BLOB_MAGIC_SIZE     (CURL_BLOB_MAGIC_STRLEN + 1)
+#define CURL_BLOB_OFFSET_DUPFLAG (CURL_BLOB_MAGIC_SIZE)
+#define CURL_BLOB_SIZE_DUPFLAG   (1)
+#define CURL_BLOB_OFFSET_DATALEN (CURL_BLOB_OFFSET_DUPFLAG + \
+                                  CURL_BLOB_SIZE_DUPFLAG)
+#define CURL_BLOB_SIZE_DATALEN   (sizeof(size_t))
+#define CURL_BLOB_OFFSET_DATAPTR (CURL_BLOB_OFFSET_DATALEN + \
+                                  CURL_BLOB_SIZE_DATALEN)
+#define CURL_BLOB_SIZE_DATAPTR   (sizeof(void *))
+#define CURL_BLOB_SIZE           (CURL_BLOB_OFFSET_DATAPTR + \
+                                  CURL_BLOB_SIZE_DATAPTR)
+
+#define CURL_BLOB_DUPFLAG_COPY   (1)
+#define CURL_BLOB_DUPFLAG_NOCOPY (0)
+
+/* the struct curl_blob store binary data parameters
+ * the structure size is 0x20 and can be followed by binary data */
+struct curl_blob {
+  char blob_internal[CURL_BLOB_SIZE];
```

Comment:
Why isn't this a proper struct instead? Perhaps like this:
~~~
struct curl_blob {
  unsigned int magic; /* must be CURL_BLOB_MAGIC */
  size_t len;
  void *data;
  bool dup;
};
~~~
As this is a `*_BLOB` option we can just require that all such options provide a pointer to such a struct, and if we use the magic field, there should be a limited risk that the wrong kind of data is passed in without detection.

*maybe* we should make the `dup` an unsigned char instead and call it `flags` so there's some teeny extension possibility for the bits that will just be unused anyway...

Reply:
" limited risk "
I want 0% risk of anyone malicious which use "curl_easy_setopt(curl, CURLOPT_SSLCERT", canot create confusion by using the magic.
So Curl_setstropt first check the string content and fail if it is the magic.

Ok to use a struct

Reply:
All the code will be so much cleaner and nicer when using a proper struct there.

Reply:
I commited this
struct curl_blob {
    unsigned char magic[CURL_BLOB_MAGIC_SIZE]; /* must be CURL_BLOB_MAGIC */
    unsigned char flags;
    void* data;
    size_t len;
};